consensus/src/liveness/proposal_generator.rs

```diff
@@ -463,23 +512,57 @@ impl ProposalGenerator {
             PIPELINE_BACKPRESSURE_ON_PROPOSAL_TRIGGERED.observe(0.0);
         };
 
+        let mut execution_backpressure_applied = false;
+        if let Some(config) = &self.pipeline_backpressure_config.execution {
+            if pipeline_pending_latency.as_millis()
+                > config.back_pressure_pipeline_latency_limit_ms as u128
+            {
+                let execution_backpressure = self
+                    .pipeline_backpressure_config
+                    .get_execution_block_size_backoff(
+                        &self
+                            .block_store
+                            .get_recent_block_execution_times(config.num_blocks_to_look_at),
+                    );
+                if let Some(execution_backpressure_block_size) = execution_backpressure {
+                    values_max_block_txns.push(
+                        (execution_backpressure_block_size as f64
+                            * config.reordering_ovarpacking_factor.max(1.0))
+                            as u64,
+                    );
+                    values_max_txns_from_block_to_execute.push(execution_backpressure_block_size);
+                    execution_backpressure_applied = true;
+                }
+            }
+        }
+        EXECUTION_BACKPRESSURE_ON_PROPOSAL_TRIGGERED.observe(
```

Comment:
do we need both this and pipeline backpressure based on timestamp? aren't they achieving similar goal?

Reply:
this drastically reduces when pipeline backpressure kicks in. but:

- this only considers execution. if commit or another stage becomes a bottleneck, we need to fall back to pipeline backpressure
- this is only local. pipeline backpressure sees if building commit cert is delayed, and kicks in - even if current node is processing thing very fast